mailet/standard/src/main/java/org/apache/james/transport/mailets/StripAttachment.java

```diff
@@ -221,122 +223,136 @@ public void service(Mail mail) throws MailetException {
-                    @SuppressWarnings("unchecked")
-                    Map<String, byte[]> m = (Map<String, byte[]>) mail.getAttribute(attributeName);
-                    if (m == null) {
-                        m = new LinkedHashMap<String, byte[]>();
-                        mail.setAttribute(attributeName, (LinkedHashMap<String, byte[]>) m);
-                    }
-                    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-                    OutputStream os = new BufferedOutputStream(
-                            byteArrayOutputStream);
-                    part.writeTo(os);
-                    m.put(fileName, byteArrayOutputStream.toByteArray());
-                }
-                if (removeAttachments.equals(REMOVE_MATCHED)) {
-                    ret = true;
+        boolean shouldRemove = removeAttachments.equals(REMOVE_ALL);
+        if (fileNameMatches(fileName)) {
+            if (directoryName != null) {
+                String filename = saveAttachmentToFile(bodyPart, Optional.fromNullable(fileName));
+                if (filename != null) {
+                    addFilenameToAttribute(mail, filename, SAVED_ATTACHMENTS_ATTRIBUTE_KEY);
                 }
             }
-            if (!ret) {
-                ret = removeAttachments.equals(REMOVE_ALL);
+            if (attributeName != null) {
+                addPartContent(bodyPart, mail, fileName);
             }
-            if (ret) {
-                @SuppressWarnings("unchecked")
-                Collection<String> c = (Collection<String>) mail
-                        .getAttribute(REMOVED_ATTACHMENTS_ATTRIBUTE_KEY);
-                if (c == null) {
-                    c = new ArrayList<String>();
-                    mail.setAttribute(REMOVED_ATTACHMENTS_ATTRIBUTE_KEY,
-                            (ArrayList<String>) c);
-                }
-                c.add(fileName);
+            if (removeAttachments.equals(REMOVE_MATCHED)) {
+                shouldRemove = true;
             }
         }
-        return ret;
+        if (shouldRemove) {
+            addFilenameToAttribute(mail, fileName, REMOVED_ATTACHMENTS_ATTRIBUTE_KEY);
+        }
+        return shouldRemove;
+    }
+
+    private void addFilenameToAttribute(Mail mail, String filename, String attributeName) {
+        @SuppressWarnings("unchecked")
+        Collection<String> attributeValues = (Collection<String>) mail.getAttribute(attributeName);
+        if (attributeValues == null) {
+            attributeValues = new ArrayList<String>();
+            mail.setAttribute(attributeName, (ArrayList<String>) attributeValues);
+        }
+        attributeValues.add(filename);
+    }
+
+    private void addPartContent(BodyPart bodyPart, Mail mail, String fileName) throws IOException, MessagingException {
+        @SuppressWarnings("unchecked")
+        Map<String, byte[]> fileNamesToPartContent = (Map<String, byte[]>) mail.getAttribute(attributeName);
```

Comment:
Then you are calling setAttribute in each cases, i'm not sure about the impacts of this.